src/Parsers/tests/gtest_format_hiliting.cpp

```diff
@@ -0,0 +1,303 @@
+#include <unordered_set>
+
+#include <Parsers/IAST.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/parseQuery.h>
+#include <gtest/gtest.h>
+#include <Common/StackTrace.h>
+
+std::string hilite(const std::string & s, const char * hilite_type)
+{
+    std::stringstream ss;
+    ss << hilite_type;
+    ss << s;
+    ss << DB::IAST::hilite_none;
+    return ss.str();
+}
+
+std::string keyword(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_keyword);
+}
+
+std::string identifier(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_identifier);
+}
+
+std::string alias(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_alias);
+}
+
+std::string op(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_operator);
+}
+
+std::string function(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_function);
+}
+
+std::string substitution(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_substitution);
+}
+
+std::vector<const char *> HILITES =
+    {
+        DB::IAST::hilite_keyword,
+        DB::IAST::hilite_identifier,
+        DB::IAST::hilite_function,
+        DB::IAST::hilite_operator,
+        DB::IAST::hilite_alias,
+        DB::IAST::hilite_substitution,
+        DB::IAST::hilite_none
+    };
+
+[[maybe_unused]] const char * consume_hilites(const char ** it)
+{
+    const char * last_hilite = nullptr;
+    while (true)
+    {
+        bool changed_hilite = false;
+        for (const char * hilite : HILITES)
+        {
+            if (std::string_view(*it).starts_with(hilite))
```

Comment:
(Unrelated fun fact and just FYI: all of this magic works only because [since C++11](https://en.cppreference.com/w/cpp/string/basic_string) the internal data of a string (l. 216) is guaranteed to be NULL-terminated.)

Reply:
Yay, some good design!